File: Canvas/PortedClasses/FPPlayer.js
// Copyright 2010 Filip Kunc. All rights reserved.

var playerImage = new Image();
playerImage.src = "Images/ball.png";

var jumpImage = new Image();
jumpImage.src = "Images/speed.png";

const tolerance = 3.0;
const maxSpeed = 5.8;
const speedPowerUp = 1.5;
const upSpeed = 7.0;
const maxFallSpeed = -15.0;
const acceleration = 1.1;
const deceleration = 1.1 * 0.2;
const changeDirectionSpeed = 3.0;
const maxSpeedUpCount = 60 * 6; // 60 FPS * 6 sec

function FPPlayer()
{
    this.x = 480.0 / 2.0 - 32.0 / 2.0;
    this.y = 320.0 / 2.0 - 32.0 / 2.0;
    this.moveX = 0.0;
    this.moveY = 0.0;
    this.rotation = 0.0;
    this.jumping = false;
    this.speedUpCounter = 0;
    this.alpha = 1.0;
    this.isVisible = true;
    
    this.rect = function()
    {
        return new FPRect(this.x, this.y, 32.0, 32.0);
    }
    
    this.move = function(offsetX, offsetY)
    {
        this.x += offsetX;
        this.y += offsetY;
    }
    
    this.update = function(game)
    {
        var inputAcceleration = game.inputAcceleration;
    	var moveLeftOrRight = false;
    	
    	if (this.speedUpCounter > 0)
    	{
    	    if (++this.speedUpCounter > maxSpeedUpCount)
    	    {
    	        this.speedUpCounter = 0;
	        }
	    }
	    
	    var currentMaxSpeed = this.speedUpCounter > 0 ? maxSpeed * speedPowerUp : maxSpeed;
	    
    	if (inputAcceleration.x < 0.0)
    	{
    		if (this.moveX < 0.0)
    			this.moveX += Math.abs(inputAcceleration.x) * acceleration * changeDirectionSpeed;
    		if (this.moveX < maxSpeed)
    			this.moveX += Math.abs(inputAcceleration.x) * acceleration;
    		moveLeftOrRight = true;
    	}
    	else if (inputAcceleration.x > 0.0)
    	{
    		if (this.moveX > 0.0)
    			this.moveX -= Math.abs(inputAcceleration.x) * acceleration * changeDirectionSpeed;
    		if (this.moveX > -maxSpeed)
    			this.moveX -= Math.abs(inputAcceleration.x) * acceleration;
    		moveLeftOrRight = true;
    	}
    	if (!this.jumping && inputAcceleration.y > 0.0)
    	{
    		if (this.moveY < upSpeed)
    			this.moveY = upSpeed;
    		this.jumping = true;
    	}

    	if (!moveLeftOrRight)
    	{
    		if (Math.abs(this.moveX) < deceleration)
    			this.moveX = 0.0;
    		else if (this.moveX > 0.0)
    			this.moveX -= deceleration;
    		else if (this.moveX < 0.0)
    			this.moveX += deceleration;
    	}	

    	this.moveY -= deceleration;
    	if (this.moveY < maxFallSpeed)
    		this.moveY = maxFallSpeed;
    	this.jumping = true;
    	
    	game.moveWorld(this.moveX, 0.0);
    	if (this.collisionLeftRight(game))
    		this.moveX = 0.0;
    	game.moveWorld(0.0, this.moveY);
    	this.collisionUpDown(game);
    	this.rotation -= this.moveX * 3.0;
    	
    	this.alpha += 0.07;
    	if (this.alpha > 3.1415926535)
    	    this.alpha -= 3.1415926535;
    }
    
    this.collisionLeftRight = function(game)
    {
    	var isColliding = false;

    	for (i in game.gameObjects)
    	{
    	    var platform = game.gameObjects[i];
    	    if (platform.isPlatform())
    		{
    			var intersection = FPRectIntersection(platform.rect(), this.rect());
    			if (intersection.isEmptyWithTolerance())
    			    continue;

    			if (platform.rect().left() > this.rect().left())
    			{
    			    if (platform.isMovable())
    			    {
    			        platform.move(intersection.size.width, 0.0);
    			        if (platform.collisionLeftRight(game))
    			        {
    			            platform.move(-intersection.size.width, 0.0);
    			            game.moveWorld(intersection.size.width, 0.0);
    			            isColliding = true;
			            }
			        }
			        else
			        {
			            game.moveWorld(intersection.size.width, 0.0);
			            isColliding = true;
		            }
    			}
    			else if (platform.rect().right() < this.rect().right())
    			{
    				if (platform.isMovable())
    			    {
    			        platform.move(-intersection.size.width, 0.0);
    			        if (platform.collisionLeftRight(game))
    			        {
    			            platform.move(intersection.size.width, 0.0);
    			            game.moveWorld(-intersection.size.width, 0.0);
    			            isColliding = true;
			            }
			        }
			        else
			        {
			            game.moveWorld(-intersection.size.width, 0.0);
			            isColliding = true;
		            }
    			}
    		}
    	}

    	return isColliding;
    }

    this.collisionUpDown = function(game)
    {
    	var isColliding = false;

    	for (i in game.gameObjects)
    	{
    	    var platform = game.gameObjects[i];
    		if (platform.isPlatform())
    		{
    			var intersection = FPRectIntersection(platform.rect(), this.rect());
    			if (intersection.isEmptyWithTolerance())
    				continue;

    			if (platform.rect().bottom() < this.rect().bottom())
    			{
    				if (this.moveY > 0.0)
    					this.moveY = 0.0;
    			
    				game.moveWorld(0.0, -intersection.size.height);
    				isColliding = true;
    			}
    			else if (this.moveY < 0.0)
    			{
    				if (platform.rect().top() > this.rect().bottom() - tolerance + this.moveY)
    				{
    					this.moveY = 0.0;
    					this.jumping = false;
    					game.moveWorld(0.0, intersection.size.height);
    					isColliding = true;
    				}
    			}
    			else if (platform.rect().top() > this.rect().bottom() - tolerance)
    			{
    				this.jumping = false;
    				game.moveWorld(0.0, intersection.size.height);
    				isColliding = true;
    			}
    		}
    	}

    	return isColliding;
    }    
    
    this.draw = function(context)
    {
        context.save();
        context.translate(240, 160);
        context.rotate(this.rotation * Math.PI / 180.0);
        context.drawImage(playerImage, -16, -16);
        context.restore();
        
        if (this.speedUpCounter > 0)
            this.drawSpeedUp(context);
    }
    
    this.drawSpeedUp = function(context)
    {
        context.globalAlpha = Math.abs(Math.sin(this.alpha)) * 0.5 + 0.5;
        context.drawImage(jumpImage, 240 - 32, 160 - 32);
        context.globalAlpha = 1.0;
    }
}

File: SharedClasses/FPPlayer.m
//
//  FPPlayer.m
//  IronJump
//
//  Created by Filip Kunc on 5/2/10.
//  For license see LICENSE.TXT
//

#import "FPPlayer.h"
#import "FPTexture.h"
#import "FPTextureArray.h"

#if TARGET_OS_MAC
FPTexture *playerTexture = nil;
FPTexture *jumpTexture = nil;
#endif

const float tolerance = 3.0f;
const float maxSpeed = 5.8f;
const float speedPowerUp = 1.5f;
const float upSpeed = 7.0f;
const float maxFallSpeed = -15.0f;
const float acceleration = 1.1f;
const float deceleration = 1.1f * 0.2f;
const float changeDirectionSpeed = 3.0f;
const int maxSpeedUpCount = 60 * 6; // 60 FPS * 6 sec

@implementation FPPlayer

@synthesize x, y, moveX, moveY, speedUpCounter, isVisible, rotation, alpha;

#if TARGET_OS_MAC
+ (FPTexture *)loadTextureIfNeeded
{
	if (!playerTexture)
	{
		playerTexture = [[FPTexture alloc] initWithFile:@"ball.png" convertToAlpha:NO];
		jumpTexture = [[FPTexture alloc] initWithFile:@"speed.png" convertToAlpha:NO];
	}
	return playerTexture;
}
#endif

- (id)init
{
	self = [super init];
	if (self)
	{
		x = 0.0f;
		y = 0.0f;
		moveX = 0.0f;
		moveY = 0.0f;
		jumping = NO;
		speedUpCounter = 0;
		alpha = 1.0f;
        isVisible = YES;
	}
	return self;
}

- (id)initWithWidth:(float)aWidth height:(float)aHeight
{
	self = [self init];
	if (self)
	{
		x = aWidth / 2.0f - 32.0f / 2.0f;
		y = aHeight / 2.0f - 32.0f / 2.0f;
	}
	return self;
}

- (void)dealloc
{
	[super dealloc];
}

- (void)encodeWithCoder:(NSCoder *)aCoder
{
	[aCoder encodeFloat:x forKey:@"x"];
	[aCoder encodeFloat:y forKey:@"y"];
}

- (id)initWithCoder:(NSCoder *)aDecoder
{
	self = [self init];
	if (self)
	{
		x = [aDecoder decodeFloatForKey:@"x"];
		y = [aDecoder decodeFloatForKey:@"y"];
	}
	return self;
}

- (CGRect)rect
{
	return CGRectMake(x, y, 32.0f, 32.0f);
}

- (BOOL)isPlatform
{
	return NO;
}

- (BOOL)isMovable
{
	return NO;
}

- (BOOL)isTransparent
{
	return YES;
}

- (void)moveWithX:(float)offsetX y:(float)offsetY
{
	x += offsetX;
	y += offsetY;
}

- (void)updateWithGame:(id<FPGameProtocol>)game
{	
	CGPoint inputAcceleration = [game inputAcceleration];
	BOOL moveLeftOrRight = NO;
	
	if (speedUpCounter > 0)
	{
		if (++speedUpCounter > maxSpeedUpCount)
		{
			speedUpCounter = 0;			
		}
	}
	
	float currentMaxSpeed = speedUpCounter > 0 ? maxSpeed * speedPowerUp : maxSpeed;
	
	if (inputAcceleration.x < 0.0f)
	{
		if (moveX < 0.0f)
			moveX += fabsf(inputAcceleration.x) * acceleration * changeDirectionSpeed;
		moveX += fabsf(inputAcceleration.x) * acceleration;
		if (moveX > currentMaxSpeed)
			moveX = currentMaxSpeed;
		
		moveLeftOrRight = YES;
	}
	else if (inputAcceleration.x > 0.0f)
	{
		if (moveX > 0.0f)
			moveX -= fabsf(inputAcceleration.x) * acceleration * changeDirectionSpeed;
		moveX -= fabsf(inputAcceleration.x) * acceleration;
		if (moveX < -currentMaxSpeed)
			moveX = -currentMaxSpeed;
		
		moveLeftOrRight = YES;
	}
	
	if (!jumping && inputAcceleration.y > 0.0f)
	{
		if (moveY < upSpeed)
			moveY = upSpeed;
		jumping = YES;
	}
	
	if (!moveLeftOrRight)
	{
		if (fabsf(moveX) < deceleration)
			moveX = 0.0f;
		else if (moveX > 0.0f)
			moveX -= deceleration;
		else if (moveX < 0.0f)
			moveX += deceleration;
	}	
	
	moveY -= deceleration;
	if (moveY < maxFallSpeed)
		moveY = maxFallSpeed;
	jumping = YES;
	
	[game moveWorldWithX:moveX y:0.0f];
	if ([self collisionLeftRight:game])
		moveX = 0.0f;
    [game moveWorldWithX:0.0f y:moveY];
	[self collisionUpDown:game];
	rotation -= moveX * 3.0f;
	
	alpha += 0.07f;
	if (alpha > M_PI)
		alpha -= M_PI;
}

- (BOOL)collisionLeftRight:(id<FPGameProtocol>)game
{
	BOOL isColliding = NO;
	
	for (id<FPGameObject> platform in [game gameObjects])
	{
		if (platform.isPlatform)
		{
			CGRect intersection = CGRectIntersection(platform.rect, self.rect);
			if (CGRectIsEmptyWithTolerance(intersection))
				continue;
			
			if (CGRectGetMinX(platform.rect) > CGRectGetMinX(self.rect))
			{
                if (platform.isMovable)
                {
                    [platform moveWithX:intersection.size.width y:0.0f];
                    if ([platform collisionLeftRight:game])
                    {
                        [platform moveWithX:-intersection.size.width y:0.0f];
                        [game moveWorldWithX:intersection.size.width y:0.0f];
                        isColliding = YES;
                    }
                }
                else
                {
                    [game moveWorldWithX:intersection.size.width y:0.0f];
                    isColliding = YES;
                }					
			}
			else if (CGRectGetMaxX(platform.rect) < CGRectGetMaxX(self.rect))
			{
                if (platform.isMovable)
                {
                    [platform moveWithX:-intersection.size.width y:0.0f];
                    if ([platform collisionLeftRight:game])
                    {
                        [platform moveWithX:intersection.size.width y:0.0f];
                        [game moveWorldWithX:-intersection.size.width y:0.0f];
                        isColliding = YES;
                    }
                }
                else
                {
                    [game moveWorldWithX:-intersection.size.width y:0.0f];
                    isColliding = YES;
                }
            }
		}
	}
	
	return isColliding;
}

- (BOOL)collisionUpDown:(id<FPGameProtocol>)game
{
	BOOL isColliding = NO;
	
	for (id<FPGameObject> platform in [game gameObjects])
	{
		if (platform.isPlatform)
		{
			CGRect intersection = CGRectIntersection(platform.rect, self.rect);
			if (CGRectIsEmptyWithTolerance(intersection))
				continue;
			
			if (CGRectGetMaxY(platform.rect) < CGRectGetMaxY(self.rect))
			{
				if (moveY > 0.0f)
					moveY = 0.0f;

				[game moveWorldWithX:0.0f y:-intersection.size.height];
				isColliding = YES;
			}
			else if (moveY < 0.0f)
			{
				if (CGRectGetMinY(platform.rect) > CGRectGetMaxY(self.rect) - tolerance + moveY)
				{
					moveY = 0.0f;
					jumping = NO;
					[game moveWorldWithX:0.0f y:intersection.size.height];
					isColliding = YES;
				}
			}
			else if (CGRectGetMinY(platform.rect) > CGRectGetMaxY(self.rect) - tolerance + moveY)
			{
				jumping = NO;
				[game moveWorldWithX:0.0f y:intersection.size.height];
				isColliding = YES;
			}
		}
	}
	
	return isColliding;
}

- (void)draw
{
#if TARGET_OS_IPHONE
	[[FPGameAtlas sharedAtlas] addPlayerAtPoint:CGPointMake(x, y) rotation:rotation];
#else
	[FPPlayer loadTextureIfNeeded];
	glPushMatrix();
	glTranslatef(x, y, 0.0f);	
	glTranslatef(playerTexture.width / 2.0f, playerTexture.height / 2.0f, 0.0f);
	glRotatef(rotation, 0, 0, 1);	
	glTranslatef(-playerTexture.width / 2.0f, -playerTexture.height / 2.0f, 0.0f);
	[playerTexture draw];
	glPopMatrix();
#endif
}

- (void)drawSpeedUp
{
	if (speedUpCounter <= 0)
		return;	
	
	glColor4f(1, 1, 1, fabsf(sinf(alpha)) * 0.5f + 0.5f);
	CGPoint point = CGPointMake(x - 16.0f, y - 16.0f);
	
#if TARGET_OS_IPHONE
	[[FPGameAtlas sharedAtlas] addSpeedEffectAtPoint:point];
	[[FPGameAtlas sharedAtlas] drawAllTiles];
#else
	[jumpTexture drawAtPoint:point];
#endif
}

- (id<FPGameObject>)duplicateWithOffsetX:(float)offsetX offsetY:(float)offsetY
{
	FPPlayer *duplicated = [[[FPPlayer alloc] init] autorelease];
	[duplicated moveWithX:x + offsetX y:y + offsetY];
	return duplicated;
}

- (void)parseXMLElement:(NSString *)elementName value:(NSString *)value
{
    if ([elementName isEqualToString:@"x"])
        x = [value floatValue];
    else if ([elementName isEqualToString:@"y"])
        y = [value floatValue];
}

- (void)writeToXML:(FPXMLWriter *)writer
{
    [writer writeElementWithName:@"x" floatValue:x];
    [writer writeElementWithName:@"y" floatValue:y];
}


@end
